core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -82,28 +81,25 @@ public TableScan appendsAfter(long newFromSnapshotId) {
   public CloseableIterable<FileScanTask> planFiles() {
     //TODO publish an incremental appends scan event
     List<Snapshot> snapshots = snapshotsWithin(table(), fromSnapshotId, toSnapshotId);
-    Iterable<CloseableIterable<FileScanTask>> files = Iterables.transform(snapshots, this::planFiles);
-    return CloseableIterable.concat(files);
-  }
-
-  private CloseableIterable<FileScanTask> planFiles(Snapshot snapshot) {
-    Predicate<ManifestFile> matchingManifests = manifest -> manifest.snapshotId() == snapshot.snapshotId();
-
-    Predicate<ManifestEntry> matchingManifestEntries =
-        manifestEntry ->
-            manifestEntry.snapshotId() == snapshot.snapshotId() &&
-            manifestEntry.status() == ManifestEntry.Status.ADDED;
-
-    ManifestGroup manifestGroup = new ManifestGroup(tableOps().io(), snapshot.manifests())
+    Set<Long> snapshotIds = Sets.newHashSet(Iterables.transform(snapshots, Snapshot::snapshotId));
+    Set<ManifestFile> manifests = FluentIterable
+        .from(snapshots)
+        .transformAndConcat(s -> s.manifests())
+        .filter(manifestFile -> snapshotIds.contains(manifestFile.snapshotId()))
```

Comment:
`manifestFileFilter` is moved out of `ManifestGroup` since now it helps us make the check of whether to use a parallelIterator much more accurate